### PR TITLE
Closes #1317: Add update hook to delete problematic legacy global footer menu blocks.

### LINKED ISSUE #1317: Enabling Global Footer module produces errors, breaks structure > menu
## Problem/Motivation
Attempting to enable Global Footer (GF) module on a migrated site (azs-president)

### Describe the bug
When GF module is installed, error is generated saying that menu blocks that GF module is trying to install already exist.  These menu blocks are: 

block.block.az_barrio_footer_menu_info
block.block.az_barrio_footer_menu_main
block.block.az_barrio_footer_menu_resources
block.block.az_barrio_footer_menu_social_media
block.block.az_barrio_footer_menu_topics

Sure enough, every site (migrated, or not), already have these blocks installed. 

After, GF Module shows installed, but /admin/structure/menu is broken. 

For example, I just enabled (2/15) the GF footer on this site: https://test-azs-covid19.pantheonsite.io/admin/structure/menu

### To Reproduce

Install GF Module on any site.

Note, before install, that all footer blocks already exist at /admin/structure/menu

## Proposed resolution

Above my pay grade.

Screen Shot 2022-02-15 at 9 36 21 AM
Screen Shot 2022-02-15 at 9 26 19 AM

### PR DESCRIPTION
**Pull request by joeparsons**
## Description
Adds update hook for deleting old menu block configs that prevent Quickstart Global Footer module from being installed.

We need to discuss the fact that this probably needs to a be a 2.1.x change only but that it adds a DB update that may screw up future DB update numbers in 2.0.x.

## Related Issue
Closes #1317 

## How Has This Been Tested?
Tested locally in lando.

### Steps to reproduce problem
- Clone az_quickstart
- Checkout the latest 2.0.x release tag (`2.0.18`) with git
```
git checkout 2.0.18
```
- Perform these steps with lando (the `config-distro-update` step is crucial)
```
lando start
lando install
lando drush config-distro-update
lando db-export
```
- Checkout main branch with git
- Perform these steps with lando
```
lando rebuild
lando install
lando db-import <export-file-name>
lando drush cr -y
lando drush updatedb
lando drush config-distro-update
```
- Try installing Quickstart Global Footer module -- see errors

### Test fix
- Checkout this PR branch
- Perform these steps with lando
```
lando rebuild
lando install
lando db-import <export-file-name>
lando drush cr -y
lando drush updatedb
lando drush config-distro-update
```
- Try installing Quickstart Global Footer module -- no errors

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I've added this Pull Request to the AZ Quickstart Github project and set it to "Review in progress".
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by trackleft — DISMISSED** [review]
Looks fine to me, hopefully nobody is using these blocks without the global footer.
We could consider deleting them only on install of the global footer if people are using them.

**Comment by joeparsons** [comment]
> Looks fine to me, hopefully nobody is using these blocks without the global footer. We could consider deleting them only on install of the global footer if people are using them.

I had the same thought so I initially tried implementing this in `az_global_footer_install()` but unfortunately that hook doesn't seem to fire early enough in the install / config-import process to work (installation still fails with the same errors).

**Comment by joeparsons** [comment]
I guess another alternative that might be worth exploring would be an event subscriber in the `az_global_footer` module that subscribes to config import events and does something about the problematic block configs then...

**Comment by joeparsons** [comment]
Suggestion from @tadean:  Only delete menu blocks if config hash matches the hash of the default config values in `az_barrio/config/optional`

**Comment by joeparsons** [comment]
Suggestion from @trackleft:  We may be able to use [`hook_module_preinstall()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_module_preinstall/9.2.x) in the `az_global_footer` module.

**Comment by joeparsons** [comment]
## 2022-02-24 Updates
- I tried using [`hook_module_preinstall()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_module_preinstall/9.2.x) but unfortunately that doesn't work because Drupal core's 
[`ModuleInstaller::install()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21ModuleInstaller.php/function/ModuleInstaller%3A%3Ainstall/9.2.x) method calls [`ConfigInstaller::checkConfigurationToInstall()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Config%21ConfigInstaller.php/function/ConfigInstaller%3A%3AcheckConfigurationToInstall/9.2.x) before invoking `hook_module_preinstall()` and `hook_install()` implementations  and [`ConfigInstaller::checkConfigurationToInstall()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Config%21ConfigInstaller.php/function/ConfigInstaller%3A%3AcheckConfigurationToInstall/9.2.x) is the method that throws the [`PreExistingConfigException`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Config%21PreExistingConfigException.php/class/PreExistingConfigException/9.2.x) if config objects already exist 😞 .
- I found checking that the `default_config_hash` of the menu blocks still matched their default state before deleting them does seem to work, so I've updated this PR to include that check.  It's worth noting that the `default_config_hash` values of those blocks have changed a few times since they were first added in [Quickstart 2.0.0-beta5](https://github.com/az-digital/az_quickstart/releases/tag/2.0.0-beta5).  This means that if someone were to try to update from an earlier version of Qucikstart 2.0.x (e.g. 2.0.7) to a Quickstart 2.1.x or 2.2.x release, this update wouldn't work and they'd still need to manually delete the footer menu block configs to be able to install the Quickstart global footer module.
- I also added a little bit more verbose messaging to the `return` lines of the update hook.

### Considerations for merging
I'd be comfortable merging this as-is now if others are also comfortable with it.  These are some things that should be considered before merging:
- This PR is really only intended for `2.1.x` (and `2.2.x`).  I do _not_ think we should backport this to `2.0.x` because the footer menu blocks are still included in `az_barrio`'s `config/install` folder in the `2.0.x` branch.  If we did want to backport this change to `2.0.x`, I think we'd need a `2.0.x` specific PR that also removes those files.
- This is blocked by #1350 and also contains the changes from #1351 so it should really be merged after that PR.  Because this branch is based on that one, this PR is currently targeting the `issue/1350` branch (to simplify review) but **should be updated to target `main` before merging**.

File: az_quickstart.install
Comment (trackleft) [inline az_quickstart.install:196]: It could be confused, but don't you need to unset the ones that are different from the $footer_menu_blocks array before you delete them all?

Comment (trackleft) [reply az_quickstart.install:196]: I suppose the install will break on any existing active config. NVM

File: az_quickstart.install
Comment (trackleft) [inline az_quickstart.install:203]: We should add the drush command to the readme in case someone needs it along with some explanation about when, if ever, they need to run it.

**Comment by trackleft** [comment]
image

**Review by trackleft — DISMISSED** [review]

**Review by tadean — DISMISSED** [review]

**Comment by joeparsons** [comment]
## 2022-03-02 Updates

- Adjusted update hook number to follow updated numbering convention (changed `92102` to `920102`).
- Added documentation about the known module installation issues to the README for the Quickstart global footer module.

**Review by bberndt-uaz — APPROVED** [review]
Tested locally as described and confirmed that the fix allows installing the Quickstart Global Footer module without errors.

**Review by tadean — APPROVED** [review]

**Review by trackleft — APPROVED** [review]

**Pull request merged**
